File: 2021/day7/day7.py
def read_file(filename):
    with open(filename, "r") as f:
        data = f.read().strip().split(",")
    return list(map(int, data))

t_positions = read_file("test_input.txt")
positions = read_file("input.txt")


# PART 1
def part1(positions):
    left = min(positions)
    right = max(positions)
    fuels = []
    for i in range(left, right+1):
        fuels.append(sum(abs(pos-i) for pos in positions))

    return min(fuels)

assert part1(t_positions) == 37
print(part1(positions))


# PART 2 Using Arithmetic Progression
def gauss(x):
    return x * (x+1) / 2

def part2(positions):
    left = min(positions)
    right = max(positions)
    fuels = []
    for i in range(left, right+1):
        fuel = 0
        for pos in positions:
            diff = abs(pos-i)
            fuel += gauss(diff)
        fuels.append(int(fuel))
    return min(fuels)

assert part2(t_positions) == 168
print(part2(positions))

File: 2021/day7/day7.go
package main

import (
    "fmt"
    "advent-of-code/tools"
    "math"
    "sort"
)

// Find Median
func part1(numbers []int) int {
    nums := make([]int, len(numbers))
    copy(nums, numbers)
    sort.Ints(nums)
    median := nums[len(nums) / 2]

    fuel := 0
    for _, n := range nums {
        fuel += int(math.Abs(float64(n - median)))
    }
    return fuel
}

func part2(numbers []int) int {
    left := tools.Min(numbers)
    right := tools.Max(numbers)
    var fuels []int
    for i := left; i <= right; i++ {
        fuel := 0
        for _, n := range numbers {
            diff := int(math.Abs(float64(n - i)))
            fuel += ar_prog(diff)
        }
        fuels = append(fuels, fuel)
    }
    return tools.Min(fuels)
}

// Arithmetic Progression sum
func ar_prog(x int) int {
    return x * (x + 1) / 2
}

func main() {
    numbers := tools.Read("input.txt", ",")
    fmt.Println(part1(numbers))
    fmt.Println(part2(numbers))

}
